go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,947 @@
+
+	reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), re.valueType)
+	*out = exec.ExecResult{
+		Type:   reeType,
+		Len:    re.inputLen,
+		Nulls:  0,
+		Offset: 0,
+		Children: []exec.ArraySpan{
+			{
+				Type: reeType.RunEnds(),
+				Len:  numOutput,
+			},
+			{
+				Type: reeType.Encoded(),
+				Len:  numOutput,
+			},
+		},
+	}
+
+	out.Children[0].Buffers[1].WrapBuffer(runEndsBuffer)
+	if validityBuffer != nil {
+		out.Children[1].Buffers[0].WrapBuffer(validityBuffer)
+	}
+	out.Children[1].Buffers[1].WrapBuffer(offsetsBuffer)
+	out.Children[1].Buffers[2].WrapBuffer(valueBuffer)
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) WriteEncodedRuns(out *exec.ExecResult) int64 {
+	outputRunEnds := exec.GetData[R](out.Children[0].Buffers[1].Buf)
+	outputOffsets := exec.GetSpanOffsets[O](&out.Children[1], 1)
+	outputValues := out.Children[1].Buffers[2].Buf
+
+	var readOffset int64
+	currentRun, curRunValid := re.readValue(readOffset)
+	readOffset++
+
+	var writeOffset, valueOffset int64
+	validityBuf := out.Children[1].Buffers[0].Buf
+	setValidity := func(valid bool) {}
+	if len(validityBuf) > 0 {
+		setValidity = func(valid bool) {
+			bitutil.SetBitTo(validityBuf, int(writeOffset), valid)
+		}
+	}
+
+	outputOffsets[0], outputOffsets = 0, outputOffsets[1:]
+
+	writeValue := func(valid bool, value []byte) {
+		setValidity(valid)
+		valueOffset += int64(copy(outputValues[valueOffset:], value))
+		outputOffsets[writeOffset] = O(valueOffset)
+	}
+
+	for readOffset < re.inputLen {
+		value, valid := re.readValue(readOffset)
+
+		if valid != curRunValid || !bytes.Equal(value, currentRun) {
+			writeValue(curRunValid, currentRun)
+			runEnd := R(readOffset)
+			outputRunEnds[writeOffset] = runEnd
+			writeOffset++
+			curRunValid, currentRun = valid, value
+		}
+		readOffset++
+	}
+
+	writeValue(curRunValid, currentRun)
+	outputRunEnds[writeOffset] = R(re.inputLen)
+	return writeOffset + 1
+}
+
+func validateRunEndType[R RunEndsType](length int64) error {
+	runEndMax := MaxOf[R]()
+	if length > int64(runEndMax) {
+		return fmt.Errorf("%w: cannot run-end encode arrays with more elements than the run end type can hold: %d",
+			arrow.ErrInvalid, runEndMax)
+	}
+	return nil
+}
+
+func createEncoder[R RunEndsType, V exec.FixedWidthTypes](input *exec.ArraySpan) *runEndEncodeLoopFixedWidth[R, V] {
+	return &runEndEncodeLoopFixedWidth[R, V]{
+		inputLen:      input.Len,
+		inputOffset:   input.Offset,
+		inputValidity: input.Buffers[0].Buf,
+		inputValues:   input.Buffers[1].Buf,
+		valueType:     input.Type,
+		readValue:     readFixedWidthVal[V],
+		writeValue:    writeFixedWidthVal[V],
+	}
+}
+
+func createVarBinaryEncoder[R RunEndsType, O int32 | int64](input *exec.ArraySpan) *runEndEncodeLoopBinary[R, O] {
+	return &runEndEncodeLoopBinary[R, O]{
+		inputLen:      input.Len,
+		inputOffset:   input.Offset,
+		inputValidity: input.Buffers[0].Buf,
+		inputValues:   input.Buffers[2].Buf,
+		offsetValues:  exec.GetSpanOffsets[O](input, 1),
```

Comment:
Does this apply `input.Offset`? That makes it a bit too different from the other buffers pointers in this struct.

Reply:
one complexity of binary arrays is that the offset values are always absolute offsets into the raw data. So in order for everything to work correctly, the slice for the raw values (`inputValues`) can't be shifted by the offset. In addition, you can't easily shift the validity bitmap in a predictable way because it's a bitmap and you aren't guaranteed the offset is easily divisible by 8. So those can't utilize the offset built in.

However, `exec.GetSpanOffsets` already accounts for the offset when creating the typed slice of offsets allowing the simple change of just not needing to apply the offset to the `readOffset` in the loop for var binary arrays.

Reply:
the same is also true for `exec.GetSpanValues`

The difference between `GetSpanValues` and `GetSpanOffsets` is that `GetSpanOffsets` adds the expected +1 to the length that offsets should have. so `len(exec.GetSpanOffsets(....)) == span.Len+1`

Also, `exec.GetSpanOffsets` doesn't do the `nil` check that `GetSpanValues` does.